GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -16,32 +19,68 @@ public class RevisionGraph : IRevisionGraphRowProvider
+
+        /// <summary>
+        /// The number of known revisions, to be nodes in the graph.
+        /// "Cache" for <see cref="_revisionByObjectId"/> count to prevent concurrent dictionary access.
+        /// </summary>
+        private int _revisionCount;
+
+        /// <summary>
+        /// Partial revisions that are seen as parents and have no GitRevision yet.
+        /// These revisions are only accessed from the update thread and are normally empty
+        /// when the loading is completed.
+        /// </summary>
+        private readonly Dictionary<ObjectId, RevisionGraphRevision> _incompleteRevisionByObjectId = new();
 
+        /// <summary>
+        /// The loading is completed when all revisions are loaded from the log,
+        /// then <see cref="_revisionByObjectId"/> is completely updated.
+        /// </summary>
         private bool _loadingCompleted;
 
         /// <summary>
         /// The max score is used to keep a chronological order during the graph building.
-        /// It is cheaper than doing <c>_nodes.Max(n => n.Score)</c>.
+        /// It is cheaper than doing <c>_nodeByObjectId.Values.Max(n => n.Score)</c>.
         /// </summary>
         private int _maxScore;
 
         /// <summary>
-        /// The node cache is an ordered list with the nodes.
+        /// The node cache is an ordered list of <see cref="_revisionByObjectId"/>.
         /// This is used so we can draw commits before the graph building is complete.
         /// </summary>
-        /// <remarks>This cache is very cheap to build.</remarks>
-        private RevisionGraphRevision[]? _orderedNodesCache;
-        private bool _reorder = true;
+        /// <remarks>
+        /// This cache is normally built twice, first time when a few revisions are loaded
+        /// and the first page is to be displayed, very quick.
+        /// The second time is when all revisions are loaded, that may require hundred ms to build.
+        /// </remarks>
+        private ImmutableArray<RevisionGraphRevision> _orderedNodesCache = ImmutableArray<RevisionGraphRevision>.Empty;
+
+        /// <summary>
+        /// Mark the <see cref="_orderedNodesCache"/> as dirty. This will trigger a rebuild of the cache.
+        /// </summary>
+        private bool _reorderNodeCache = true;
+
+        /// <summary>
+        /// The score of the last node in <see cref="_orderedNodesCache"/>.
+        /// </summary>
         private int _orderedUntilScore = -1;
 
         /// <summary>
         /// The ordered row cache contains rows with segments stored in lanes.
         /// </summary>
-        /// <remarks>This cache is very expensive to build.</remarks>
-        private IList<RevisionGraphRow>? _orderedRowCache;
+        /// <remarks>This cache is very expensive to build, done in chunks to limit load.</remarks>
+        private IList<RevisionGraphRow> _orderedRowCache = new List<RevisionGraphRow>();
```

Comment:
```suggestion
        private List<RevisionGraphRow> _orderedRowCache = new();
```
This also saves a call of `.ToList()` below.

Reply:
This was changed to ImmutableArray